### PR TITLE
Attribute Summary Consistency Changes

### PR DESCRIPTION
**Pull request by sboptimal**
I have been through and modified all attributes to use consistent wording for their XML summaries.  There were a couple of grammatical changes in there so it is worth checking all files to see if you are happy with the wording.

### CODE REVIEW

**Comment by sboptimal** [comment]
Not sure why the AppVeyor build is failing.  Looking at the error I don't think it is related to my pull request.  If it is let me know and I will fix.

**Comment by CharliePoole** [comment]
Appreciated as your work is, I'm not in favor of a chef to the same attribute that is being documented. It takes space and doesn't seem to have any benefit for the user.

**Comment by Sbosanquet** [comment]
Hi @CharliePoole, in an earlier correction to an incorrect comment it was requested that I add it in for that one. From a consistency point of view they should either all have them or not. It's easy enough to remove the cref and attribute name completely from the summary if that's what is needed.  Either way, I feel they need to all be consistent.

**Comment by CharliePoole** [comment]
Sorry, I haven't been following along here but just happened to glance in. When you have a `cref` to the same class you are documenting, what do you see when you hover? What when you click? Does it seem useful?

As the retiring member, I'd say pay more attention to other folks comments than mine, but please do think this through in terms of what the user sees.

**Comment by CharliePoole** [comment]
So it is useful at the point of reference but not at the point of definition. I never thought of it that way, which is why I hadn't generally use it. I imagine the definition also appears when hovering over a list of options provided by Intellisense, so I can see the fix makes sense.

Thanks for the example!

**Comment by ChrisMaddock** [comment]
Just FYI - this will need a review from a second team member - so you may prefer to wait for both sets of feedback before making any further changes. 🙂

**Comment by sboptimal** [comment]
@ChrisMaddock ok, not a problem.  I have corrected those you highlighted.  Amazing how they slip through.  If you want to check the TimeoutAttribute description I have removed the reference to a new thread as requested.

File: src/NUnitFramework/framework/Attributes/ApartmentAttribute.cs [hunk 1]
```diff
@@ -29,7 +29,7 @@
 namespace NUnit.Framework
 {
     /// <summary>
-    /// Marks a test that must run in a particular threading apartment state, causing it
+    /// <see cref="ApartmentAttribute"/> marks a test that must run in a particular threading apartment state, causing it
```
Comment (jnm2) [inline src/NUnitFramework/framework/Attributes/ApartmentAttribute.cs]: While I would use crefs any time a code element is named, I wouldn't name the attribute here. [The BCL doesn't do this](https://github.com/dotnet/corefx/blob/d4541b21dcb0cb15c5ec49c38354bde46e3eaea4/src/System.ComponentModel.Primitives/src/System/ComponentModel/DescriptionAttribute.cs#L7-L11) and the reason that occurs to me is that the name is already right above the text, so starting off the sentence with the same name is visually redundant:

image

I know it's confusing that I didn't ask to entirely remove the existing content in https://github.com/nunit/nunit/pull/2952, but it did cross my mind. :D I love that you're shooting for consistency. That's something we could really use. If you're willing to help us in this direction, I'd prefer to start the sentence with the class name being an implicit subject noun like I see in other projects.

Comment (jnm2) [reply src/NUnitFramework/framework/Attributes/ApartmentAttribute.cs]: Checking with @nunit/framework-team, is view you'd share?

Comment (sboptimal) [reply src/NUnitFramework/framework/Attributes/ApartmentAttribute.cs]: @jnm2 not a problem.  Lets get it sorted.  As I said to Charlie, I'm willing to redo them so they are all have the same structure and consistency.  It isn't a problem.  I'll wait until the general consensus is back and then we can go from there.

Comment (ChrisMaddock) [reply src/NUnitFramework/framework/Attributes/ApartmentAttribute.cs]: > Checking with @nunit/framework-team, is this a view you'd share?

I'm apathetic. Happy to accept this, as @sboptimal's put the work in to tidy things up, but I have no strong opinion on whether we're consistent one way, or the other. 🙂

Comment (jnm2) [reply src/NUnitFramework/framework/Attributes/ApartmentAttribute.cs]: Okay :D Just to clarify, there's a ton in this PR that I want to merge whichever way we go. I'm proposing that in addition to these changes (and as a reversal of only some of them) we delete the attribute names as the first words of the summaries and start with the next word.

Comment (ChrisMaddock) [reply src/NUnitFramework/framework/Attributes/ApartmentAttribute.cs]: Sounds good to me!

Comment (mikkelbu) [reply src/NUnitFramework/framework/Attributes/ApartmentAttribute.cs]: I don't feel strongly about this. If @sboptimal doesn't minding doing the additional changes then I don't see any reason not to do this. Otherwise, I can also live with the current solution.

Comment (sboptimal) [reply src/NUnitFramework/framework/Attributes/ApartmentAttribute.cs]: Ok, so just to confirm:
1. The Cref at the start of each summary will be removed along with the attribute name
1. All attributes will now start with something like: Marks..., Applies..., Supplies... etc.  Basically, whatever their first word is after the attribute name.

Old:
```
    /// <summary>
    /// <see cref="RandomAttribute"/> supplies a set of random values
    /// to a single parameter of a parameterized test.
    /// </summary>
```

New:
```
    /// <summary>
    /// Supplies a set of random values to a single parameter of a parameterized test.
    /// </summary>
```

Looking at the BCL link you sent through earlier.  It seems they wrap the summary comments in a <para> tag for formatting reasons I guess.  Do you want your attributes doing the same way as i'll do them at the same time if so?

Comment (jnm2) [reply src/NUnitFramework/framework/Attributes/ApartmentAttribute.cs]: @sboptimal That's what I had in mind, yes. We don't want to use `<para>` tags line the BCL did in that example, though. We're targeting only IntelliSense with our XML docs and we only use `<para>` to wrap at least two paragraphs.

Comment (sboptimal) [reply src/NUnitFramework/framework/Attributes/ApartmentAttribute.cs]: Ok, not a problem.  I'll try to get it done this week.

**Comment by sboptimal** [comment]
Sorry for the delay.  So as discussed last week here is the new batch of changes, Have a read through and check your are happy.  I'll correct any that need changing.

File: src/NUnitFramework/framework/Attributes/MaxTimeAttribute.cs [hunk 2]
```diff
@@ -29,7 +29,7 @@
 namespace NUnit.Framework
 {
     /// <summary>
-    /// Specifies the maximum time (in milliseconds) for a test case to succeed.
```
Comment (ChrisMaddock) [inline src/NUnitFramework/framework/Attributes/MaxTimeAttribute.cs]: Why this change, sorry?

File: src/NUnitFramework/framework/Attributes/OneTimeSetUpAttribute.cs [hunk 3]
```diff
@@ -26,8 +26,8 @@ namespace NUnit.Framework
 	using System;
 
 	/// <summary>
-	/// Attribute used to identify a method that is called once
-	/// to perform setup before any child tests are run.
+	/// Identifies a method to be called once, prior to the child tests executing, 
+	/// to perform any setup requirements for the proceding tests.
```
Comment (ChrisMaddock) [inline src/NUnitFramework/framework/Attributes/OneTimeSetUpAttribute.cs]: *preceding - but I think this is the wrong word anyway. Maybe you mean 'descendant' tests - but is that perhaps overcomplicating the explanation? Can you clarify what you wanted to change about the previous description here, and maybe we can come up with a better summary?

File: src/NUnitFramework/framework/Attributes/NonTestAssemblyAttribute.cs [hunk 4]
```diff
@@ -27,12 +27,11 @@
...
 {
     /// <summary>
-    /// The NonTestAssemblyAttribute may be used by third-party frameworks
-    /// or other software that references the NUnit framework but does not
-    /// contain tests. Applying the attribute indicates that the assembly
-    /// is not a test assembly and may prevent errors if certain runners
-    /// attempt to load the assembly. Note that recognition of the attribute
-    /// depends on each individual runner.
+    /// Used by third-party frameworks, or other software, that reference 
+    /// the NUnit framework but does not contain any tests. Applying the
```
Comment (ChrisMaddock) [inline src/NUnitFramework/framework/Attributes/NonTestAssemblyAttribute.cs]: does -> do

File: src/NUnitFramework/framework/Attributes/RepeatAttribute.cs [hunk 5]
```diff
@@ -36,8 +36,7 @@
 namespace NUnit.Framework
 {
     /// <summary>
-    /// RepeatAttribute may be applied to test case in order
-    /// to run it multiple times.
+    /// Specified that a test should be run multiple times.
```
Comment (ChrisMaddock) [inline src/NUnitFramework/framework/Attributes/RepeatAttribute.cs]: *Specifies

File: src/NUnitFramework/framework/Attributes/SingleThreadedAttribute.cs [hunk 6]
```diff
@@ -28,11 +28,11 @@
...
-    /// SingleThreadedAttribute applies to a test fixture and indicates
-    /// that all the child tests must be run on the same thread as the
-    /// OneTimeSetUp and OneTimeTearDown. It sets a flag in the
-    /// TestExecutionContext and forces all tests to be run sequentially
-    /// on the current thread. Any ParallelScope setting is ignored.
+    /// Marks a test fixture as requiring all child tests must be run on the 
+    /// same thread as the OneTimeSetUp and OneTimeTearDown. A flag in the
+    /// <see cref="TestExecutionContext"/> is set and forces all tests to be run 
+    /// sequentially on the current thread. Any <see cref="ParallelScope"/> 
+    /// setting is ignored.
```
Comment (ChrisMaddock) [inline src/NUnitFramework/framework/Attributes/SingleThreadedAttribute.cs]: "A flag in the TestExecutionContext is set and forces..." doesn't sound right. Should this be "which forces" now, instead?

File: src/NUnitFramework/framework/Attributes/TestCaseAttribute.cs [hunk 7]
```diff
@@ -32,8 +32,7 @@
 namespace NUnit.Framework
 {
     /// <summary>
-    /// TestCaseAttribute is used to mark parameterized test cases
-    /// and provide them with their arguments.
+    /// Marks the method as a parameterized test cases and provide them with their arguments.
```
Comment (ChrisMaddock) [inline src/NUnitFramework/framework/Attributes/TestCaseAttribute.cs]: Pluralization problems in this one. ('Marks' vs 'provide')

File: src/NUnitFramework/framework/Attributes/TestCaseSourceAttribute.cs [hunk 8]
```diff
@@ -33,8 +33,7 @@
 namespace NUnit.Framework
 {
     /// <summary>
-    /// TestCaseSourceAttribute indicates the source to be used to
-    /// provide test cases for a test method.
+    /// Identifies the source to be used as the test cases for a test method.
```
Comment (ChrisMaddock) [inline src/NUnitFramework/framework/Attributes/TestCaseSourceAttribute.cs]: Not sure about the re-wording here - the source and the test cases are different things, really. Can we switch this one back to the old version? `Identifies the source to be used to provide test cases for a test method.`

File: src/NUnitFramework/framework/Attributes/TimeoutAttribute.cs [hunk 9]
```diff
@@ -30,10 +30,10 @@
...
 {
     /// <summary>
-    /// Used on a method, marks the test with a timeout value in milliseconds. 
-    /// The test will be run in a separate thread and is cancelled if the timeout 
-    /// is exceeded. Used on a class or assembly, sets the default timeout 
-    /// for all contained test methods.
+    /// Applies a timeout to a test. When used on a method it marks the test 
+    /// with a timeout value in milliseconds. The test is cancelled if the 
+    /// timeout is exceeded. When used on a class or assembly it sets the 
+    /// default timeout for all contained test methods.
```
Comment (ChrisMaddock) [inline src/NUnitFramework/framework/Attributes/TimeoutAttribute.cs]: I see how you've tried to reword here - the original in this case was ambiguous, I think. The timeout is always in milliseconds, whether the attribute is on a method/class/assembly.

**Comment by sboptimal** [comment]
@ChrisMaddock I have been through and made some changes.  Hopefully should make it easier for the other guys to review if you are happy with one set.  I have rolled back some of the wording as requested.  Others I have corrected as per your suggestions.  If you can have a quick scan when you get chance.

File: src/NUnitFramework/framework/Attributes/TimeoutAttribute.cs [hunk 10]
```diff
@@ -30,10 +30,9 @@
 namespace NUnit.Framework
 {
     /// <summary>
-    /// Used on a method, marks the test with a timeout value in milliseconds. 
-    /// The test will be run in a separate thread and is cancelled if the timeout 
-    /// is exceeded. Used on a class or assembly, sets the default timeout 
-    /// for all contained test methods.
+    /// Applies a timeout in milliseconds to a test. 
+    /// When applied to a method the test is cancelled if the timeout is exceeded. 
+    /// When applied to a class or assembly it sets the default timeout for all contained test methods.
```
Comment (jnm2) [inline src/NUnitFramework/framework/Attributes/TimeoutAttribute.cs]: There should be a comma after each “When” phrase.

File: src/NUnitFramework/framework/Attributes/CultureAttribute.cs [hunk 11]
```diff
@@ -29,8 +29,7 @@
 namespace NUnit.Framework
 {
     /// <summary>
-    /// CultureAttribute is used to mark a test fixture or an
-    /// individual method as applying to a particular Culture only.
+    /// Marks a test fixture or method as applying to a specific Culture.
```
Comment (mikkelbu) [inline src/NUnitFramework/framework/Attributes/CultureAttribute.cs]: Should this also mention assembly, since the attribute has `AttributeTargets.Assembly`?

File: src/NUnitFramework/framework/Attributes/DatapointSourceAttribute.cs [hunk 12]
```diff
@@ -27,11 +27,10 @@
...
 {
     /// <summary>
-    /// Used to mark a field, property or method providing a set of datapoints to 
-    /// be used in executing any theories within the same fixture that require an 
-    /// argument of the Type provided. The data source may provide an array of
-    /// the required Type or an <see cref="IEnumerable{T}"/>.
-    /// Synonymous with DatapointsAttribute.
+    /// Marks a field, property or method as providing a set of datapoints for use 
+    /// in executing any theories within the same fixture that require an argument 
+    /// of the Type provided. The data source may provide an array of the required
```
Comment (mikkelbu) [inline src/NUnitFramework/framework/Attributes/DatapointSourceAttribute.cs]: For `DatapointsAttribute` we have `... of the provided Type. ...` and not `...of the Type provided. ...` as here. I prefer the `DatapointsAttribute` variant.

File: src/NUnitFramework/framework/Attributes/DescriptionAttribute.cs [hunk 13]
```diff
@@ -27,8 +27,7 @@
 namespace NUnit.Framework
 {
     /// <summary>
-    /// Attribute used to provide descriptive text about a 
-    /// test case or fixture.
+    /// Provides the descriptive text relating to the test case or fixture.
```
Comment (mikkelbu) [inline src/NUnitFramework/framework/Attributes/DescriptionAttribute.cs]: What about assembly?

File: src/NUnitFramework/framework/Attributes/ExplicitAttribute.cs [hunk 14]
```diff
@@ -28,10 +28,9 @@
 namespace NUnit.Framework
 {
     /// <summary>
-    /// ExplicitAttribute marks a test or test fixture so that it will
-    /// only be run if explicitly executed from the GUI or command line
-    /// or if it is included by use of a filter. The test will not be
-    /// run simply because an enclosing suite is run.
+    /// Marks a test or test fixture such that it will only run if explicitly
```
Comment (mikkelbu) [inline src/NUnitFramework/framework/Attributes/ExplicitAttribute.cs]: What about assembly?

File: src/NUnitFramework/framework/Attributes/IgnoreAttribute.cs [hunk 15]
```diff
@@ -29,9 +29,7 @@
 namespace NUnit.Framework
 {
     /// <summary>
-    /// Attribute used to mark a test that is to be ignored.
-    /// Ignored tests result in a warning message when the
-    /// tests are run.
+    /// Marks a test as being ignored. Ignored tests result in a warning message when the tests are run.
```
Comment (mikkelbu) [inline src/NUnitFramework/framework/Attributes/IgnoreAttribute.cs]: The other places we have also mentioned test fixture (and perhaps also assembly)

File: src/NUnitFramework/framework/Attributes/PlatformAttribute.cs [hunk 16]
```diff
@@ -29,8 +29,7 @@
 namespace NUnit.Framework
 {
     /// <summary>
-    /// PlatformAttribute is used to mark a test fixture or an
-    /// individual method as applying to a particular platform only.
+    /// Marks a test fixture or method as applying to a specific platform.
```
Comment (mikkelbu) [inline src/NUnitFramework/framework/Attributes/PlatformAttribute.cs]: Assembly?

File: src/NUnitFramework/framework/Attributes/SetCultureAttribute.cs [hunk 17]
```diff
@@ -28,11 +28,9 @@
 namespace NUnit.Framework
 {
     /// <summary>
-    /// Sets the current Culture for the duration of a test.
-    /// <para>
-    /// It may be specified at the level of a test or a fixture.
-    /// The culture remains set until the test or fixture completes and is then reset to its original value.
-    /// </para>
+    /// Sets the current Culture for the duration of a test. The culture remains set until 
+    /// the test or fixture completes and is then reset to its original value.
```
Comment (mikkelbu) [inline src/NUnitFramework/framework/Attributes/SetCultureAttribute.cs]: Assembly?

File: src/NUnitFramework/framework/Attributes/SetUICultureAttribute.cs [hunk 18]
```diff
@@ -28,11 +28,9 @@
 namespace NUnit.Framework
 {
     /// <summary>
-    /// Sets the current UI Culture for the duration of a test.
-    /// <para>
-    /// It may be specified at the level of a test or a fixture.
-    /// The UI culture remains set until the test or fixture completes and is then reset to its original value.
-    /// </para>
+    /// Sets the current UI Culture for the duration of a test. The UI culture remains set until
+    /// the test or fixture completes and is then reset to its original value.
```
Comment (mikkelbu) [inline src/NUnitFramework/framework/Attributes/SetUICultureAttribute.cs]: Assembly?

File: src/NUnitFramework/framework/Attributes/SetUpAttribute.cs [hunk 19]
```diff
@@ -26,8 +26,7 @@ namespace NUnit.Framework
     using System;
 
     /// <summary>
-    /// Attribute used to identify a method that is called 
-    /// immediately before each test is run.
+    /// Identifies a method to be called prior to the running of each test.
```
Comment (mikkelbu) [inline src/NUnitFramework/framework/Attributes/SetUpAttribute.cs]: I prefer "Identifies a method to be called immediately before each test is run." to match the TearDownAttribute.

File: src/NUnitFramework/framework/Attributes/TestOfAttribute.cs [hunk 20]
```diff
@@ -33,7 +33,7 @@
 namespace NUnit.Framework
 {
     /// <summary>
-    /// Indicates which class the test or test fixture is testing
+    /// Indicates the method or class the test or test fixture is testing
```
Comment (mikkelbu) [inline src/NUnitFramework/framework/Attributes/TestOfAttribute.cs]: Assembly? We should also end the sentence with a comma.

Comment (jnm2) [reply src/NUnitFramework/framework/Attributes/TestOfAttribute.cs]: @mikkelbu You meant, end the sentence with a period?

Comment (mikkelbu) [reply src/NUnitFramework/framework/Attributes/TestOfAttribute.cs]: Yes :)
